R/dependencies.R

```diff
@@ -0,0 +1,410 @@
+##'                  through the tree
+##'
+##' @export
+orderly_build_dep_tree <- function(name, id = "latest", root = NULL,
+                                   locate = TRUE, upstream = FALSE, con = NULL,
+                                   propagate = FALSE, max_depth = 100,
+                                   list_all = FALSE) {
+  if (is.null(con)) {
+    con <- orderly_db("destination", orderly_config_get(root, locate))
+    on.exit(DBI::dbDisconnect(con))
+  }
+
+  dep_tree <- build_tree(name = name, id = id, depth = max_depth, con = con,
+                         upstream = upstream, list_all = list_all)
+
+  # propagate out-of-date
+  if (propagate) {
+    propagate(dep_tree$root, upstream)
+  }
+
+  dep_tree
+}
+
+propagate <- function(vertex, upstream) {
+  for (child in vertex$children) {
+    if (!upstream) {
+      if (vertex$out_of_date) {
+        child$out_of_date = TRUE
+      }
+    }
+
+    propagate(child, upstream)
+
+    if (upstream) {
+      if (child$out_of_date) {
+        vertex$out_of_date = TRUE
+      }
+    }
+  }
+}
+
+Vertex <- R6::R6Class("Vertex", list(
+  parent = NULL,
+  children = list(),
+  name = NULL,
+  id = NULL,
+  out_of_date = NULL,
+  initialize = function(parent, name, id, out_of_date) {
+    # add some type safety here
+    self$parent <- parent
+    self$name <- name
+    self$id <- id
+    self$out_of_date <- out_of_date
+  },
+  add_child = function(child) {
+    self$children <- append(self$children, list(child))
+  },
+  format = function() {
+    return(sprintf("%s [%s]", self$name, self$id))
+  }
+  )
+)
+
+
+Tree <- R6::R6Class("Tree", list(
+  vertices = NULL,
+  root = NULL,
+  message = NULL,
+  initialize = function(root) {
+    self$vertices <- append(self$vertices, list(root))
+    self$root <- root
+    self$message <- NULL
+  },
+  add_child = function(parent, name, id, out_of_date) {
+    child <- Vertex$new(parent, name, id, out_of_date)
+    parent$add_child(child)
+    self$vertices <- append(self$vertices, list(child))
+    return(child)
+  },
+  # this is stupidly hacky to get the formatting right
+  format_helper = function(vertex = self$root, fvector = c(), str="") {
+    ## if the tree has a warning message, append it to the front in the red
+    if (!is.null(self$message)) {
+      str <- paste(str, crayon::red(self$message), "\n", collapse = "")
+      self$message <- NULL
+    }
+
+    console_colour <- if (vertex$out_of_date) {crayon::red} else {crayon::blue}
+
+    if (length(fvector) == 0) {
+      str <- paste(str, console_colour(sprintf("%s", vertex$format())), "\n", collapse = "")
+    } else {
+      spacing <- paste(ifelse(head(fvector, -1), "| ", "  "), collapse = "")
+      str <- paste(str, console_colour(sprintf("%s|___%s", spacing, vertex$format())), "\n", collapse = "")
+    }
+
+    if (length(vertex$children) > 0) {
+      for (i in 1:length(vertex$children)) {
+        child <- vertex$children[[i]]
+
+        if (i != length(vertex$children)) {
```

Comment:
this mgiht be simpler as

```
thing <- i != length(vertex$children)
str <- self$format_helper(child, c(fvector, thing), str)
```

though I am unclear what the second argument here represents at this point